File: projects/22-lsm-redis/src/block_cache.rs
//! V7 — Block cache: a hand-built LRU over decoded SSTable blocks. `src/block_cache.rs`.
//!
//! An SSTable (V4) stores its data as fixed-ish **blocks** (a few KiB each). A point
//! read locates the one block that could hold the key, reads it from disk, and searches
//! within it. Under a skewed workload (a hot working set — Zipfian, like every real
//! cache) the *same* blocks are read over and over. Reading and decoding them from disk
//! every time is the read-amplification tax the LSM shape imposes; the **block cache**
//! is how you pay it once.
//!
//! It's a bounded, in-memory map from block identity → decoded block, with **LRU**
//! eviction: on a hit, mark the block most-recently-used; on an insert past the byte
//! budget, evict least-recently-used blocks until you fit. This is the one cache you
//! build *by hand* (no `cargo add lru`) because doing so is the point — an intrusive
//! `HashMap` + doubly-linked list (or a map + a recency queue) that stays O(1) per op.
//! Bound it by **bytes**, not entries: blocks vary in size and the budget is memory.
//!
//! Concurrency: many connections read at once, so the state is behind a lock (or shard
//! it to cut contention — note which in `docs/22-design.md`). Blocks are handed out as
//! `Arc<Block>` so a reader holds one cheaply while others touch the cache.
//!
//! *Concept to internalize:* how a block cache bounds an LSM's read amplification, why
//! LRU approximates "keep the working set" (recency ≈ reuse), and the classic O(1) LRU
//! data structure — plus how bounding by bytes (not count) matches a real memory budget.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use bytes::Bytes;

/// Identifies one block within the whole store: which SSTable, and its byte offset
/// inside that file. Cheap to hash/compare — the cache key.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BlockKey {
    pub sstable_id: u64,
    pub offset: u64,
}

/// A decoded data block, ready to search. Kept as raw bytes here; a real impl might
/// cache the parsed key/value offsets so a hit skips re-parsing too.
pub type Block = Bytes;

/// State guarded by the lock. Split out so the V7 TODO is about the *algorithm*
/// (recency + eviction), not the locking.
#[derive(Default)]
struct Inner {
    map: HashMap<BlockKey, Arc<Block>>,
    /// Sum of cached block sizes in bytes — kept `<= cap_bytes`.
    used_bytes: usize,
    // TODO(V7): add whatever recency bookkeeping your LRU needs (e.g. a doubly-linked
    // list of keys, or a monotonic access counter per entry). This is the part that
    // makes eviction O(1) and "least-recently-used" actually true.
    hits: u64,
    misses: u64,
}

/// A bounded LRU cache of decoded SSTable blocks. `new`/`stats` are wired; the get /
/// insert / eviction algorithm is V7.
pub struct BlockCache {
    cap_bytes: usize,
    inner: Mutex<Inner>,
}

impl BlockCache {
    /// A cache holding at most `cap_bytes` of blocks. `cap_bytes == 0` disables caching
    /// (every `get` misses, every `insert` is a no-op) — the read path then goes
    /// straight to disk, which is the default until you build this.
    pub fn new(cap_bytes: usize) -> Self {
        BlockCache {
            cap_bytes,
            inner: Mutex::new(Inner::default()),
        }
    }

    pub fn capacity_bytes(&self) -> usize {
        self.cap_bytes
    }

    /// `(hits, misses)` since start — the source of the block-cache hit-ratio metric the
    /// boss fight grades. Wired.
    pub fn stats(&self) -> (u64, u64) {
        let g = self.inner.lock().expect("block cache lock poisoned");
        (g.hits, g.misses)
    }

    /// Look up a block, counting the hit/miss and (on a hit) marking it most-recently-used.
    ///
    /// TODO(V7): on a hit, bump recency and return the `Arc<Block>` clone; on a miss,
    /// record it and return `None`. When `cap_bytes == 0`, always miss. Recording
    /// hits/misses here is what feeds the hit-ratio metric.
    pub fn get(&self, key: &BlockKey) -> Option<Arc<Block>> {
        let _ = key;
        todo!("V7: LRU lookup — on hit bump recency + count hit; else count miss → None")
    }

    /// Insert a freshly-read block, evicting least-recently-used blocks to stay within
    /// the byte budget.
    ///
    /// TODO(V7): insert the block, add its size to `used_bytes`, and while
    /// `used_bytes > cap_bytes` evict the LRU block (subtracting its size) until it
    /// fits. A single block larger than `cap_bytes` should simply not be cached rather
    /// than evict everything. No-op when `cap_bytes == 0`.
    pub fn insert(&self, key: BlockKey, block: Arc<Block>) {
        let _ = (key, block);
        todo!("V7: insert + mark MRU, then evict LRU blocks until used_bytes <= cap_bytes")
    }
}

#[cfg(test)]
mod tests {
    // TODO(V7): prove bounded, correct LRU.
    //   - byte bound: after many inserts, `used_bytes` never exceeds `cap_bytes`;
    //   - LRU order: fill to capacity, touch block A, insert one more → the evicted
    //     block is the least-recently-used, NOT A;
    //   - hit accounting: a get after insert is a hit; a get for an absent key is a
    //     miss; the ratio matches the sequence you drove;
    //   - `cap_bytes == 0` disables cleanly: get always misses, insert never grows.
}

File: projects/22-lsm-redis/src/lsm_redis/block_cache.py
"""V7 — Block cache: a hand-built LRU over decoded SSTable blocks.
`src/lsm_redis/block_cache.py`.

An SSTable (V4) stores its data as blocks of a few KiB. A point read locates the
one block that could hold the key, reads it from disk, and searches inside it.
Under a skewed workload — a hot working set, Zipfian, like every real cache —
the *same* blocks are read over and over. Paying the read and the decode every
time is the read-amplification tax the LSM shape imposes; the block cache is how
you pay it once.

It is a bounded map from block identity to block, with **LRU** eviction: on a
hit, mark the block most-recently-used; on an insert past the byte budget, evict
least-recently-used blocks until you fit. Bound it by **bytes**, not entries:
blocks vary in size and the budget you actually have is memory.

*Concept to internalize:* how a block cache bounds an LSM's read amplification,
why LRU approximates "keep the working set" (recency ≈ reuse), the classic O(1)
LRU structure, and why bounding by bytes rather than count is what makes the
bound mean something.

## What "build it by hand" means in Python

The Rust version said "no `cargo add lru`". The Python translation is not "no
`OrderedDict`" — that would be a superstition rather than a lesson.

**`functools.lru_cache` is the forbidden shortcut**, and for concrete reasons:
it bounds by *entry count* (so it cannot satisfy the byte-bound criterion at
all), it exposes no eviction hook and no per-entry size, its statistics are a
`CacheInfo` tuple you cannot label by outcome, and it is keyed on function
arguments, so "the same block" and "the same call" have to be the same thing.
Every one of those is a criterion V7 grades.

**`collections.OrderedDict` is not a shortcut** — it *is* the hash map plus
intrusive doubly-linked list you would hand-roll, implemented in C and exposed
as `move_to_end(key)` and `popitem(last=False)`, both O(1). Reaching for it is
the same engineering decision as reaching for `HashMap` in Rust. What stays
yours is everything the vertical is actually about: the byte accounting, the
eviction loop, the too-big-to-cache case, the hit/miss counters, and the
disabled case. If you would rather build the linked list explicitly once to see
it, that is a genuinely good hour — but do it as a comparison, not as penance.

## The lock, which is a real decision here and not a formality

Rust needed a `Mutex` because threads. Whether *you* need one depends on a
choice you make in `engine.py`:

* If every cache operation happens on the event loop with no `await` between
  the read and the write, a run of plain Python statements is already
  indivisible — no other coroutine can observe a half-updated `used_bytes`. No
  lock, and adding an `asyncio.Lock` would cost a scheduler round-trip per
  block lookup to prevent a race that cannot happen.
* But disk reads must **not** run on the event loop, which means block reads
  land in a thread pool — and if the *insert* happens in that worker thread,
  two threads now touch this object and you need a real `threading.Lock`. The
  GIL is not that lock: `used_bytes -= size` is a read, a subtract and a store
  with bytecode boundaries in between.

Both designs work. The one that avoids the lock keeps the pool worker dumb — it
reads bytes and returns them, and the coroutine does the cache insert on the
loop. The one that takes the lock keeps the pool worker whole. Pick one on
purpose and write the reasoning in `docs/22-design.md`; the SPEC grades the
decision, not the choice.

If you do take a lock, `threading.Lock` is uncontended-cheap in CPython, and
sharding (N locks by `hash(key) % N`) is the standard next step when it is not —
also the standard premature optimization when contention was never measured.
"""

from __future__ import annotations

from collections import OrderedDict
from typing import NamedTuple

__all__ = ["Block", "BlockCache", "BlockKey", "CacheStats"]

type Block = bytes
"""A block as read from disk. `bytes` rather than `bytearray` on purpose: it is
immutable, so it can be handed to any number of concurrent readers with no copy
and no risk that one of them mutates a block another is mid-search."""


class BlockKey(NamedTuple):
    """Identifies one block in the whole store: which SSTable, and where in it.

    A `NamedTuple` because it needs to be hashable, cheap to compare, and
    readable at the call site — and because tuple hashing is C-level, which
    matters for something on the hot path of every read.
    """

    sstable_id: int
    offset: int


class CacheStats(NamedTuple):
    """Hits and misses since start — the source of the hit-ratio metric the boss
    fight grades at ≥ 95%."""

    hits: int
    misses: int


class BlockCache:
    """A byte-bounded LRU over decoded SSTable blocks.

    `capacity_bytes`, `used_bytes` and `stats` are wired; `get`, `insert` and
    the eviction that keeps the bound true are V7.
    """

    def __init__(self, capacity_bytes: int) -> None:
        self.capacity_bytes = capacity_bytes
        """Byte budget. `0` disables the cache entirely: every `get` misses and
        every `insert` is a no-op, so the read path goes straight to disk. That
        has to work — it is both a V7 criterion and the state the scaffold runs
        in before you build this."""

        self._blocks: OrderedDict[BlockKey, Block] = OrderedDict()
        """Insertion-ordered map, used as recency-ordered: oldest (least
        recently used) first, so `popitem(last=False)` is the eviction."""

        self._used_bytes = 0
        self._hits = 0
        self._misses = 0

    @property
    def used_bytes(self) -> int:
        return self._used_bytes

    def stats(self) -> CacheStats:
        return CacheStats(self._hits, self._misses)

    def __len__(self) -> int:
        return len(self._blocks)

    def get(self, key: BlockKey) -> Block | None:
        """Look up a block, counting the outcome and marking a hit as
        most-recently-used.

        TODO(V7): on a hit, bump recency (`move_to_end`) and count a hit; on a
        miss, count a miss and return `None`. When `capacity_bytes == 0`, always
        miss — and note that a disabled cache still counting misses is correct
        and useful: the ratio then honestly reads 0%.

        Counting here rather than at the call site is deliberate. The hit ratio
        must describe every lookup the read path makes, and a counter the caller
        has to remember to bump is a counter that will disagree with reality on
        exactly the code path someone added in a hurry.
        """
        raise NotImplementedError(
            "V7: LRU lookup — on hit bump recency + count hit; else count miss"
        )

    def insert(self, key: BlockKey, block: Block) -> None:
        """Insert a freshly-read block, evicting LRU blocks to stay in budget.

        TODO(V7): store the block, add `len(block)` to `_used_bytes`, then evict
        from the least-recently-used end until `_used_bytes <= capacity_bytes`.
        Three cases that a naive version gets wrong, each of which is a
        criterion:

        * **Re-inserting a key already present** must not double-count its
          bytes. Subtract the old size before adding the new one, or the bound
          drifts upward until it is not a bound.
        * **A block larger than the whole capacity** should simply not be
          cached. Inserting it first and then evicting to fit empties the entire
          cache to make room for something that cannot stay — one oversized
          block would flush your whole working set.
        * **`capacity_bytes == 0`** is a no-op, not an insert followed by an
          immediate eviction.

        The byte-bound criterion is "`used_bytes` never exceeds `capacity_bytes`,
        regardless of block sizes or insert order" — *never*, including
        transiently in the middle of this method, if anything can observe it.
        Under the no-lock design nothing can, because there is no `await` here;
        under the threaded design something can, which is one more reason the
        choice in the module docstring is real.
        """
        raise NotImplementedError(
            "V7: insert + mark MRU, then evict LRU blocks until used_bytes <= capacity"
        )
